dvc/checkout.py

```diff
@@ -169,135 +130,109 @@ def _checkout_file(
 ):
     """The file is changed we need to checkout a new copy"""
     modified = False
-    cache_info = cache.hash_to_path_info(obj.hash_info.value)
-    if fs.exists(path_info):
-        if not relink and _changed(path_info, fs, obj, cache, state=state):
-            modified = True
-            _remove(path_info, fs, cache, force=force)
-            _link(cache, cache_info, path_info)
-        else:
+    cache_info = cache.hash_to_path_info(change.new.obj.hash_info.value)
+    if change.old.obj:
+        if relink:
             if fs.iscopy(path_info) and _cache_is_copy(cache, path_info):
                 cache.unprotect(path_info)
             else:
-                _remove(path_info, fs, cache, force=force)
+                _remove(path_info, fs, change.old.in_cache, force=force)
                 _link(cache, cache_info, path_info)
+        else:
+            modified = True
+            _remove(path_info, fs, change.old.in_cache, force=force)
+            _link(cache, cache_info, path_info)
     else:
         _link(cache, cache_info, path_info)
         modified = True
 
     if state:
-        state.save(path_info, fs, obj.hash_info)
+        state.save(path_info, fs, change.new.obj.hash_info)
 
     if progress_callback:
         progress_callback(str(path_info))
 
     return modified
 
 
-def _remove_redundant_files(
-    path_info, fs, obj, cache, force, dvcignore: Optional[DvcIgnoreFilter]
+def _diff(
+    path_info,
+    fs,
+    obj,
+    cache,
+    relink=False,
+    dvcignore: Optional[DvcIgnoreFilter] = None,
 ):
-    if dvcignore:
-        existing_files = set(dvcignore.walk_files(fs, path_info))
-    else:
-        existing_files = set(fs.walk_files(path_info))
+    old = None
+    try:
+        _, old = stage(
+            cache,
+            path_info,
+            fs,
+            obj.hash_info.name if obj else cache.fs.PARAM_CHECKSUM,
+            dry_run=True,
```

Comment:
Not related to this PR, `stage` but not really looks weird. `dry_run=True` should probably be just `hash()/`hash_file()`.

Reply:
I also don't like `dry_run`, could probably consider removing it and defaulting to `True`. IIRC the only difference is that we'll use a staging reference odb, which should be fine for internals. I don't think we should use `hash/hash_file`, because those are misleading, we should not be operating on hashes, but on objects here.

Reply:
I was talking about something like [`git_odb_hashfile`](https://libgit2.org/libgit2/#HEAD/group/odb/git_odb_hashfile) or [`pygit2.hash_file`](https://www.pygit2.org/objects.html#pygit2.hashfile). Even though it has `hash` in the name, it's similar to what `stage(dry_run=True)` is doing.